text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,282 @@
+   - This key fetching latency is immaterial to the total latency of the request.
+3. How will key rotation work in a production system?
+   - Hopefully it will be a rare event.
+   - KEK rotation will require re-encrypting every DEK encrypted with the KEK (typically everything in a ~24-hour period).
+   - DEK rotation will require re-encrypting a sequence of bytes in a blob.
+
+# Extensions
+
+By extending the schema of the "file_part_byte_range" table to include a "type" column we can further reduce the number of bytes returned to the client. The client has different requirements for different sets of data. The player may only need the next `n` seconds worth of data, the console and network tabs may paginate their events, and the timeline will always fetch a simplified view of the entire recording.
+
+With the byte range pattern in place these behaviors are possible and can be exposed to the client. The ultimate outcome of this change is faster loading times and the elimination of browser freezes and crashes from large replays.
+
+This will increase the number of rows written to our database table. We would write four rows whereas with the original proposal we were only writing one. Therefore we should select our database carefully to ensure it can handle this level of write intensity.
+
+# Technical Details
+
+## Storage Service Support
+
+The following sections describe the psuedo-code necessary to fetch a range of bytes from a service provider and also links to the documentation where applicable.
+
+**Google Cloud Storage**
+
+```python
+from google.cloud.storage import Blob
+
+blob = Blob(filename, bucket)
+blob.download_as_bytes(start=start, end=stop)
+```
+
+Source: https://cloud.google.com/python/docs/reference/storage/latest/google.cloud.storage.blob.Blob#google_cloud_storage_blob_Blob_download_as_bytes
+
+**AWS S3**
+
+```python
+from boto3 import client
+
+response = client("s3", **auth).get_object(
+    Bucket=bucket,
+    Key=filename,
+    Range=f"bytes={start}-{stop}",
+)
+response["Body"].read()
+```
+
+Source: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/get_object.html
+
+**Filesystem**
+
+```python
+with open(filename, "r") as f:
+    f.seek(start)
+    f.read((stop - start) + 1)  # Range is inclusive.
+```
+
+## Consumer Buffering Mechanics
+
+The following section is highly specific to the Session Replay product.
+
+We will continue our approach of using _at-least once processing_. Each message we receive is guaranteed to be processed to completion regardless of error or interrupt. Duplicate messages are possible under this scheme and must be accounted for in the planning of each component.
+
+**Buffer Location and Behavior**
+
+The buffer is kept as an in-memory list inside the consumer process. For each message we receive we append the message to the buffer. Afterwards, we check if the buffer is full. If it is we flush. Else we wait for another message.
```

Comment:
It seems this will constrain you to a fairly small buffer size. 
Also it ties the number of replicas of your consumer to the cost efficiency of the storage, which is quite undesirable:
Assuming you are never going to commit on kafka untill the buffer is flushed (if you did you would not be able to guarantee at least once):
- If you increase the number of replicas for any reason, each replicas takes less traffic, thus it takes longer to fill the buffer.
- Unless you tie the commit batch time and size to the number of replicas (which is undesirable - see the snuba consumer), increasing the number of replicas would increase the number of files written per unit of time. 
replicas and file size should not be connected to each other, otherwise we will have to keep in mind a lot of affected moving parts when scaling the consumer.

Reply:
If the time to accumulate a batch is less than the time to upload a batch then you need to add a replica.  That's the only constraint.  You get more efficiency at peak load so its best to run our replicas hot.  The deadline will prevent the upload from sitting idle too long.  The total scale factor will be determined by the number of machines we can throw at the problem.

Multi-processing/threading, I think, will be deadly to this project.  So we will need a lot of single-threaded machines running.

I re-wrote this response several times.  Its as disorganized as my thoughts are on this.  Happy to hear critiques.

Reply:
> Unless you tie the commit batch time and size to the number of replicas (which is undesirable - see the snuba consumer), increasing the number of replicas would increase the number of files written per unit of time.

I think this is an okay outcome.  If you double the number of replicas you halve the number of parts per file and double the number of files.  That reduces cost efficiency but the throughput efficiency remains the same for each replica.  Ignoring replica count, cost efficiency will ebb and flow with the variations in load we receive throughout the day.

We still come out ahead because the total number of files written per second is less than the current implementation which is 1 file per message.